azure-identity/azure/identity/credentials/providers.py

```diff
@@ -0,0 +1,185 @@
+from azure.core.pipeline import Pipeline
+from azure.core.pipeline.policies import ContentDecodePolicy, HeadersPolicy, NetworkTraceLoggingPolicy, RetryPolicy
+from azure.core.pipeline.transport import RequestsTransport
+from msal import TokenCache
+
+from .exceptions import AuthenticationError
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import
+    from typing import Iterable, Optional, Tuple, Union
+    from azure.core import Configuration
+    from ._protocols import SupportsGetToken
+
+# pylint:disable=too-few-public-methods
+
+
+class ClientSecretTokenProvider:
+    _OAUTH_ENDPOINT = "https://login.microsoftonline.com/{}/oauth2/v2.0/token"
+
+    def __init__(self, client_id, secret, tenant_id, config=None, **kwargs):
+        # type: (str, str, str, Optional[Configuration], Mapping[str, Any]) -> None
+        if not client_id:
+            raise ValueError("client_id")
+        if not secret:
+            raise ValueError("secret")
+        if not tenant_id:
+            raise ValueError("tenant_id")
+
+        config = config or self.create_config(**kwargs)
+        # TODO: ContentDecodePolicy doesn't accept kwargs
+        policies = [ContentDecodePolicy(), config.logging_policy, config.retry_policy]
+        self._pipeline = Pipeline(transport=RequestsTransport(configuration=config), policies=policies)
+        self._cache = TokenCache()
+        self._formdata = {"client_id": client_id, "client_secret": secret, "grant_type": "client_credentials"}
+        self._auth_url = self._OAUTH_ENDPOINT.format(tenant_id)
+
+    def get_token(self, scopes):
+        # type: (Iterable[str]) -> str
+        tokens = self._cache.find(TokenCache.CredentialType.ACCESS_TOKEN, list(scopes))
+        for token in tokens:
+            if all((scope in token["target"] for scope in scopes)):
+                if int(token["expires_on"]) - 300 > int(time()):
+                    return token["secret"]
+
+        data = self._formdata.copy()
+        data["scope"] = " ".join(scopes)
+        request = HttpRequest("POST", self._auth_url, data=data)
+        response = self._pipeline.run(request, stream=False)
+        payload = response.context["deserialized_data"]
+        self._cache.add({"response": payload, "scope": scopes})
+        try:
+            return payload["access_token"]
+        except KeyError:
+            raise AuthenticationError("Unexpected response from Active Directory: {}".format(payload))
+
+    @staticmethod
+    def create_config(**kwargs):
+        config = Configuration(**kwargs)
+        config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
+        # TODO: RetryPolicy should become more straightforwardly configurable
+        config.retry_policy = RetryPolicy(**kwargs)
+        config.retry_policy._retry_on_status_codes = [404, 429] + [
+            x for x in range(500, 600)
+        ]  # pylint:disable=protected-access
+        return config
+
+
+class EnvironmentTokenProvider(ClientSecretTokenProvider):
```

Comment:
This class should eventually support both client secret auth, and certificate auth, so using ClientSecretTokenProvider as a base probably won't work long term.